File: server/sonar-ws-client/src/main/java/org/sonar/wsclient/services/Event.java
/*
 * SonarQube, open source software quality management tool.
 * Copyright (C) 2008-2014 SonarSource
 * mailto:contact AT sonarsource DOT com
 *
 * SonarQube is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * SonarQube is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
package org.sonar.wsclient.services;

import javax.annotation.CheckForNull;
import javax.annotation.Nullable;

import java.util.Date;

public class Event extends Model {

  private String id;
  private String name;
  private String category;
  private String description;
  private String resourceKey;
  private Date date;

  @CheckForNull
  public String getId() {
    return id;
  }

  public Event setId(@Nullable String id) {
    this.id = id;
    return this;
  }

  @CheckForNull
  public String getName() {
    return name;
  }

  public Event setName(@Nullable String name) {
    this.name = name;
    return this;
  }

  @CheckForNull
  public String getCategory() {
    return category;
  }

  public Event setCategory(@Nullable String category) {
    this.category = category;
    return this;
  }

  @CheckForNull
  public String getDescription() {
    return description;
  }

  public Event setDescription(@Nullable String description) {
    this.description = description;
    return this;
  }

  @CheckForNull
  public Date getDate() {
    return date;
  }

  public Event setDate(@Nullable Date date) {
    this.date = date;
    return this;
  }

  @CheckForNull
  public String getResourceKey() {
    return resourceKey;
  }

  public Event setResourceKey(@Nullable String resourceKey) {
    this.resourceKey = resourceKey;
    return this;
  }

  @Override
  public boolean equals(Object o) {
    if (this == o) {
      return true;
    }
    if (o == null || getClass() != o.getClass()) {
      return false;
    }

    Event event = (Event) o;
    return !(id != null ? !id.equals(event.id) : event.id != null);
  }

  @Override
  public int hashCode() {
    return id != null ? id.hashCode() : 0;
  }
}

File: server/sonar-web/src/main/js/apps/projectActivity/components/Event.js
/*
 * SonarQube
 * Copyright (C) 2009-2018 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
// @flow
import React from 'react';
import EventInner from './EventInner';
import ChangeEventForm from './forms/ChangeEventForm';
import RemoveEventForm from './forms/RemoveEventForm';
import { DeleteButton, EditButton } from '../../../components/ui/buttons';
/*:: import type { Event as EventType } from '../types'; */

/*::
type Props = {
  analysis: string,
  canAdmin: boolean,
  changeEvent: (event: string, name: string) => Promise<*>,
  deleteEvent: (analysis: string, event: string) => Promise<*>,
  event: EventType,
  isFirst: boolean
};
*/

/*::
type State = {
  changing: boolean,
  deleting: boolean
};
*/

export default class Event extends React.PureComponent {
  /*:: mounted: boolean; */
  /*:: props: Props; */
  state /*: State */ = {
    changing: false,
    deleting: false
  };

  componentDidMount() {
    this.mounted = true;
  }

  componentWillUnmount() {
    this.mounted = false;
  }

  startChanging = () => {
    this.setState({ changing: true });
  };

  stopChanging = () => {
    if (this.mounted) {
      this.setState({ changing: false });
    }
  };

  startDeleting = () => {
    this.setState({ deleting: true });
  };

  stopDeleting = () => {
    if (this.mounted) {
      this.setState({ deleting: false });
    }
  };

  render() {
    const { event, canAdmin } = this.props;
    const isOther = event.category === 'OTHER';
    const isVersion = !isOther && event.category === 'VERSION';
    const canChange = isOther || isVersion;
    const canDelete = isOther || (isVersion && !this.props.isFirst);
    const showActions = canAdmin && (canChange || canDelete);

    return (
      <div className="project-activity-event">
        <EventInner event={this.props.event} />

        {showActions && (
          <div className="project-activity-event-actions spacer-left">
            {canChange && (
              <EditButton className="js-change-event button-small" onClick={this.startChanging} />
            )}
            {canDelete && (
              <DeleteButton className="js-delete-event button-small" onClick={this.startDeleting} />
            )}
          </div>
        )}

        {this.state.changing && (
          <ChangeEventForm
            changeEventButtonText={
              'project_activity.' + (isVersion ? 'change_version' : 'change_custom_event')
            }
            changeEvent={this.props.changeEvent}
            event={this.props.event}
            onClose={this.stopChanging}
          />
        )}

        {this.state.deleting && (
          <RemoveEventForm
            analysis={this.props.analysis}
            deleteEvent={this.props.deleteEvent}
            event={this.props.event}
            onClose={this.stopDeleting}
            removeEventButtonText={
              'project_activity.' + (isVersion ? 'remove_version' : 'remove_custom_event')
            }
            removeEventQuestion={`project_activity.${
              isVersion ? 'remove_version' : 'remove_custom_event'
            }.question`}
          />
        )}
      </div>
    );
  }
}
